### PR TITLE
PYTHON-3729 use PyObject_GetAddr instead of PyObject_GetAddrString

### PR DESCRIPTION
**Pull request by sleepyStick**
https://jira.mongodb.org/browse/PYTHON-3729

### CODE REVIEW

File: pymongo/_cmessagemodule.c [hunk 1]
```diff
@@ -977,7 +982,12 @@ PyInit__cmessage(void)
         goto fail;
     }
 
-    GETSTATE(m)->_cbson = _cbson;
+    state = GETSTATE(m);
+    state->_cbson = _cbson;
+    state->_max_bson_size = PyUnicode_FromString("max_bson_size");
```
Comment (ShaneHarvey) [inline pymongo/_cmessagemodule.c:987]: Same comments about "_str" and error checking.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: bson/_cbsonmodule.c [hunk 2]
```diff
@@ -2137,7 +2165,7 @@ static PyObject* get_value(PyObject* self, PyObject* name, const char* buffer,
 
             /* convert to local time */
             if (options->tzinfo != Py_None) {
-                astimezone = PyObject_GetAttrString(value, "astimezone");
+                astimezone = PyObject_GetAttr(value, state->_astimezone);
                 Py_DECREF(value);
                 if (!astimezone) {
                     Py_DECREF(replace);
```
[the remark this answers is not in the document]
Comment (sleepyStick) [reply bson/_cbsonmodule.c:2174]: Done!

File: bson/_cbsonmodule.c [hunk 3]
```diff
@@ -468,8 +482,24 @@ static int _load_python_objects(PyObject* module) {
     PyObject* compiled = NULL;
     struct module_state *state = GETSTATE(module);
 
-    /* Python str for faster _type_marker check */
-    state->_type_marker_str = PyUnicode_FromString("_type_marker");
+    /* Python str for faster string checking */
```
Comment (ShaneHarvey) [inline bson/_cbsonmodule.c:485]: How about we say: 
> Cache commonly used attribute names to improve performance.

File: bson/_cbsonmodule.c [hunk 4]
```diff
@@ -617,9 +647,9 @@ int convert_codec_options(PyObject* self, PyObject* options_obj, codec_options_t
     if (type_marker < 0) {
         return 0;
     }
-
+    struct module_state *state = GETSTATE(self);
```
Comment (ShaneHarvey) [inline bson/_cbsonmodule.c:650]: Due to legacy reasons, local variables like this need to be defined at the top of the function.

File: bson/_cbsonmodule.c [hunk 5]
```diff
@@ -3067,6 +3113,21 @@ static int _cbson_clear(PyObject *m) {
     Py_CLEAR(GETSTATE(m)->UTC);
     Py_CLEAR(GETSTATE(m)->REType);
     Py_CLEAR(GETSTATE(m)->_type_marker_str);
+    Py_CLEAR(GETSTATE(m)->_type_marker_str);
```
Comment (ShaneHarvey) [inline bson/_cbsonmodule.c:3116]: _type_marker_str appears twice here.

**Review by ShaneHarvey — CHANGES_REQUESTED** [review]

File: bson/_cbsonmodule.c (hunk 3 above)
Comment (sleepyStick) [reply bson/_cbsonmodule.c]: I like that better!

File: bson/_cbsonmodule.c (hunk 5 above)
Comment (sleepyStick) [reply bson/_cbsonmodule.c]: thanks for catching that!

File: bson/_cbsonmodule.c (hunk 4 above)
Comment (sleepyStick) [reply bson/_cbsonmodule.c]: Makes sense. Changed.

File: bson/_cbsonmodule.c [hunk 6]
```diff
@@ -3051,6 +3083,21 @@ static int _cbson_traverse(PyObject *m, visitproc visit, void *arg) {
     Py_VISIT(GETSTATE(m)->MaxKey);
     Py_VISIT(GETSTATE(m)->UTC);
     Py_VISIT(GETSTATE(m)->REType);
+    Py_VISIT(GETSTATE(m)->_type_marker_str);
```
Comment (ShaneHarvey) [inline bson/_cbsonmodule.c:3086]: _type_marker_str was missed here in the initial PR. Could you open a bug ticket to fix that in 4.4.1? We'll need to create a v4.4 branch to backport this change.

**Review by ShaneHarvey — APPROVED** [review]
LGTM! Could you summarize the perf improvement for encoding regex, RawBSONDocument, Binary, Code, Timestamp, Decimal128, and datetime (decoding)?
